### PR TITLE
Add attention-mask support for ViTModel

### PR DESCRIPTION
**Pull request by Ryan-Qiyu-Jiang**
# What does this PR do?

Transformers currently only support head-masks, however recent works
in vision language modeling have used the ViTModel with attention-masks.
Adding attention-masks to ViTSelfAttention and ViTModel allow
transformers modeling_vit to be used in implementing these models.

This commit adds attention-mask implementations from BertSelfAttention
as the forward param attention_mask. This change preserves the existing
behavior by default.

A specific example of how this change will be useful is being able to
implement ViLT https://arxiv.org/abs/2102.03334, using ViTModel as the main trunk.

Unit-tests seemed to pass, though there wasn't a unit-test for head_masks,
so I just added a check for attention_mask in the signature.
Thanks!

src/transformers/models/vit/modeling_vit.py @LysandreJik, @sgugger

### CODE REVIEW

File: src/transformers/models/deit/modeling_deit.py [hunk 1]
```diff
@@ -157,6 +157,9 @@ def forward(self, hidden_states, head_mask=None, output_attentions=False):
         attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
 
         attention_scores = attention_scores / math.sqrt(self.attention_head_size)
+        if attention_mask is not None:
+            # Apply the attention mask is (precomputed for all layers in DeiTModel forward() function)
```
Comment (sgugger) [inline src/transformers/models/deit/modeling_deit.py:161]: Why the `is` here?

**Review by sgugger — COMMENTED** [review]
Thanks for the PR! I'm not sure how the API can be used however. The attention mask is just added to the attention scores, so it should already be a tensor with big negative floats for the values masked, but the other Transformer models use a mask of booleans (converted to the format of floats behind the scenes). Shouldn't that conversion also be in this PR?

**Review by patil-suraj — COMMENTED** [review]
Adding `attention_mask` as you mentioned makes sense.

But as @sgugger said, the boolean mask should be modified into a broadcasted tensor with big negative floats for the values masked. This could be done in `ViTModel.forward ` method, as is done in `BertModel`.

https://github.com/huggingface/transformers/blob/5e3b4a70d3d17f2482d50aea230f7ed42b3a8fd0/src/transformers/models/bert/modeling_bert.py#L966-L969

This `extended_attention_mask` should then be passed to the attention layer instead of the boolean mask.

(Thank you for `ViLT` reference, that looks like an interesting model :) )

File: src/transformers/models/deit/modeling_deit.py (hunk 1 above)
Comment (Ryan-Qiyu-Jiang) [reply src/transformers/models/deit/modeling_deit.py]: Fixed!

File: src/transformers/models/deit/modeling_deit.py [hunk 2]
```diff
@@ -146,7 +146,7 @@ def transpose_for_scores(self, x):
         x = x.view(*new_x_shape)
         return x.permute(0, 2, 1, 3)
 
-    def forward(self, hidden_states, head_mask=None, output_attentions=False):
+    def forward(self, hidden_states, head_mask=None, output_attentions=False, attention_mask=None):
```
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_deit.py:149]:
```suggestion
    def forward(self, hidden_states, attention_mask=None, head_mask=None, output_attentions=False):
```
For consistency with other models ;)

File: src/transformers/models/deit/modeling_deit.py [hunk 3]
```diff
@@ -226,8 +229,8 @@ def prune_heads(self, heads):
         self.attention.all_head_size = self.attention.attention_head_size * self.attention.num_attention_heads
         self.pruned_heads = self.pruned_heads.union(heads)
 
-    def forward(self, hidden_states, head_mask=None, output_attentions=False):
-        self_outputs = self.attention(hidden_states, head_mask, output_attentions)
+    def forward(self, hidden_states, head_mask=None, output_attentions=False, attention_mask=None):
+        self_outputs = self.attention(hidden_states, head_mask, output_attentions, attention_mask)
```
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_deit.py:233]:
```suggestion
        self_outputs = self.attention(hidden_states, attention_mask, head_mask, output_attentions)
```

File: src/transformers/models/deit/modeling_deit.py [hunk 4]
```diff
@@ -283,11 +286,12 @@ def __init__(self, config):
         self.layernorm_before = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
         self.layernorm_after = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
 
-    def forward(self, hidden_states, head_mask=None, output_attentions=False):
+    def forward(self, hidden_states, head_mask=None, output_attentions=False, attention_mask=None):
         self_attention_outputs = self.attention(
             self.layernorm_before(hidden_states),  # in DeiT, layernorm is applied before self-attention
             head_mask,
             output_attentions=output_attentions,
+            attention_mask=attention_mask,
```
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_deit.py:294]:
```suggestion
            self.layernorm_before(hidden_states),  # in DeiT, layernorm is applied before self-attention
            attention_mask=attention_mask,
            head_mask,
            output_attentions=output_attentions,
```

File: src/transformers/models/deit/modeling_deit.py [hunk 5]
```diff
@@ -333,6 +337,7 @@ def forward(
         output_attentions=False,
         output_hidden_states=False,
         return_dict=True,
+        attention_mask=None,
```
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_deit.py:340]: For some reason Github doesn't allow me to comment on the lines above, but one should pass the `attention_mask` after `hidden_states`.

File: src/transformers/models/deit/modeling_deit.py [hunk 6]
```diff
@@ -355,9 +360,10 @@ def custom_forward(*inputs):
                     create_custom_forward(layer_module),
                     hidden_states,
                     layer_head_mask,
+                    attention_mask,
                 )
             else:
-                layer_outputs = layer_module(hidden_states, layer_head_mask, output_attentions)
+                layer_outputs = layer_module(hidden_states, layer_head_mask, output_attentions, attention_mask)
```
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_deit.py:366]:
```suggestion
                layer_outputs = layer_module(hidden_states, attention_mask, layer_head_mask, output_attentions)
```

File: src/transformers/models/vit/modeling_vit.py [hunk 7]
```diff
@@ -180,7 +180,7 @@ def transpose_for_scores(self, x):
         x = x.view(*new_x_shape)
         return x.permute(0, 2, 1, 3)
 
-    def forward(self, hidden_states, head_mask=None, output_attentions=False):
+    def forward(self, hidden_states, head_mask=None, output_attentions=False, attention_mask=None):
```
Comment (NielsRogge) [inline src/transformers/models/vit/modeling_vit.py:183]:
```suggestion
    def forward(self, hidden_states, attention_mask=None, head_mask=None, output_attentions=False, attention_mask=None):
```

File: src/transformers/models/vit/modeling_vit.py [hunk 8]
```diff
@@ -258,8 +261,8 @@ def prune_heads(self, heads):
         self.attention.all_head_size = self.attention.attention_head_size * self.attention.num_attention_heads
         self.pruned_heads = self.pruned_heads.union(heads)
 
-    def forward(self, hidden_states, head_mask=None, output_attentions=False):
-        self_outputs = self.attention(hidden_states, head_mask, output_attentions)
+    def forward(self, hidden_states, head_mask=None, output_attentions=False, attention_mask=None):
+        self_outputs = self.attention(hidden_states, head_mask, output_attentions, attention_mask)
```
Comment (NielsRogge) [inline src/transformers/models/vit/modeling_vit.py:265]:
```suggestion
    def forward(self, hidden_states, attention_mask=None, head_mask=None, output_attentions=False):
        self_outputs = self.attention(hidden_states, attention_mask, head_mask, output_attentions)
```

File: src/transformers/models/vit/modeling_vit.py [hunk 9]
```diff
@@ -510,6 +516,7 @@ def forward(
         output_hidden_states=None,
         interpolate_pos_encoding=None,
         return_dict=None,
+        attention_mask=None,
```
Comment (NielsRogge) [inline src/transformers/models/vit/modeling_vit.py:519]: Here similarly, please pass the `attention_mask` after the `pixel_values`.

File: src/transformers/models/vit/modeling_vit.py [hunk 10]
```diff
@@ -548,12 +555,25 @@ def forward(
...
+            attention_mask, (batch_size, seq_length), device
+        )
+
         encoder_outputs = self.encoder(
             embedding_output,
             head_mask=head_mask,
             output_attentions=output_attentions,
             output_hidden_states=output_hidden_states,
             return_dict=return_dict,
+            attention_mask=extended_attention_mask,
```
Comment (NielsRogge) [inline src/transformers/models/vit/modeling_vit.py:576]: Same comment here.

Comment (NielsRogge) [inline src/transformers/models/vit/modeling_vit.py:576]:
```suggestion
        encoder_outputs = self.encoder(
            embedding_output,
            attention_mask=extended_attention_mask,
            head_mask=head_mask,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            return_dict=return_dict,
```
Same comment here.

**Review by NielsRogge — CHANGES_REQUESTED** [review]
Thanks for this PR!

I left only one comment everywhere: for consistency with other models, the `attention_mask` argument should come after the `pixel_values`/`hidden_states`.

**Comment by Ryan-Qiyu-Jiang** [comment]
@NielsRogge reordered params!

**Comment by NielsRogge** [comment]
Thanks. I think this should also be added for the Flax implementation of ViT, for consistency.

It could perhaps also be added for BEiT.

**Comment by Ryan-Qiyu-Jiang** [comment]
@NielsRogge added to flax_vit, beit, and flax_beit.

**Comment by NielsRogge** [comment]
Thanks. Can you also add the corresponding tests? In `test_modeling_vit.py` (and similarly, in test_modeling_deit.py, test_modeling_beit.py and the flax models), we can add the attention_mask to the `prepare_config_and_inputs` method, and use them in all tests.

**Comment by NielsRogge** [comment]
@Ryan-Qiyu-Jiang the shape of the `attention_mask` provided to these models is equal to (batch_size, seq_len), with the sequence length equal to the number of patches + 1 for the [CLS] token (and + 2 for DeiT, as this model also adds a special distillation token to the sequence).

This should also be added to the docstrings of the respective models.

File: src/transformers/models/beit/modeling_flax_beit.py [hunk 11]
```diff
@@ -292,6 +299,7 @@ def __call__(
             query_states,
             key_states,
             bias=attention_bias,
+            mask=attention_mask,
```
Comment (patil-suraj) [inline src/transformers/models/beit/modeling_flax_beit.py:302]: the `dot_product_attention_weights` function does not have an argument called the mask. It takes `bias` as an argument that is expanded tensor with float values. beit uses relative attention, so the `attention_mask` here should be expanded and then added to `attention_bias`. T5 does something similar which you could use here

expand the mask
https://github.com/huggingface/transformers/blob/6c08840628a22a4d53ae563d1041479649d1a8e7/src/transformers/models/t5/modeling_flax_t5.py#L403

convert it to float tensor with large negative values for masked tokens
https://github.com/huggingface/transformers/blob/6c08840628a22a4d53ae563d1041479649d1a8e7/src/transformers/models/t5/modeling_flax_t5.py#L413-L418

add it to the position bias
https://github.com/huggingface/transformers/blob/6c08840628a22a4d53ae563d1041479649d1a8e7/src/transformers/models/t5/modeling_flax_t5.py#L426-L427

File: src/transformers/models/vit/modeling_flax_vit.py [hunk 12]
```diff
@@ -171,6 +179,7 @@ def __call__(self, hidden_states, deterministic: bool = True, output_attentions:
         attn_weights = dot_product_attention_weights(
             query_states,
             key_states,
+            mask=attention_mask,
```
Comment (patil-suraj) [inline src/transformers/models/vit/modeling_flax_vit.py:182]: same comment as above, no `mask` argument. Should be expanded to a tensor and pass as bias. Here is how it's done in bert

https://github.com/huggingface/transformers/blob/6c08840628a22a4d53ae563d1041479649d1a8e7/src/transformers/models/bert/modeling_flax_bert.py#L229-L254

**Review by patil-suraj — CHANGES_REQUESTED** [review]
I left a couple of comments about how `attention_mask` should be passed in flax model. 
It should also be added in the call method of `FlaxViTPreTrainedModel` and `FlaxBeitPreTrainedModel` as is done in `FlaxBertPreTrainedModel`. This call method is the main interface for all flax models and is always called when we call any flax model.

https://github.com/huggingface/transformers/blob/6c08840628a22a4d53ae563d1041479649d1a8e7/src/transformers/models/bert/modeling_flax_bert.py#L566-L611

Let me know if you need help with this :)
